lib/lin.ml

```diff
@@ -150,45 +122,71 @@ module MakeDomThr(Spec : CmdSpec)
     | (Some i, _) :: cmds -> extract_env (i::env) cmds
     | (None,   _) :: cmds -> extract_env     env  cmds
 
+  let cmds_use_var var cmds =
+    List.exists (fun (_,c) -> Spec.cmd_uses_var var c) cmds
+
+  let ctx_doesnt_use_var _var = false
+
+  let rec shrink_cmd_list_spine cmds ctx_uses_var = match cmds with
+    | [] -> Iter.empty
+    | c::cs ->
+      let open Iter in
+      (match fst c with
+       | None -> (* shrink tail first as fewer other vars should depend on it *)
+         (map (fun cs -> c::cs) (shrink_cmd_list_spine cs ctx_uses_var)
+          <+>
+          return cs) (* not a t-assignment, so try without it *)
+       | Some i -> (* shrink tail first as fewer other vars should depend on it *)
+         (map (fun cs -> c::cs) (shrink_cmd_list_spine cs ctx_uses_var)
+         <+>
+         if cmds_use_var i cs || ctx_uses_var i
+         then empty
+         else return cs)) (* t_var not used, so try without it *)
+
+  let rec shrink_individual_cmds env cmds = match cmds with
+    | [] -> Iter.empty
+    | (opt,cmd) :: cmds ->
+      let env' = match opt with None -> env | Some i -> i::env in
+      let open Iter in
+      map (fun cmds -> (opt,cmd)::cmds) (shrink_individual_cmds env' cmds)
+      <+>
+      map (fun cmd -> (opt,cmd)::cmds) (Spec.shrink_cmd env cmd)
+
   let shrink_triple' (seq,p1,p2) =
     let open Iter in
     (* Shrinking heuristic:
        First reduce the cmd list sizes as much as possible, since the interleaving
        is most costly over long cmd lists. *)
-    let concat_map f it = flatten (map f it) in
-    let fix_seq seq =
-      let seq_env = extract_env [0] seq in
-      let triple seq p1 p2 = (seq,p1,p2) in
-      map triple (fix_cmds [0] seq) <*> fix_cmds seq_env p1 <*> fix_cmds seq_env p2
-    in
-    let seq_env = extract_env [0] seq in
-
-    concat_map fix_seq (Shrink.list_spine seq)
+    map (fun seq -> (seq,p1,p2)) (shrink_cmd_list_spine seq (fun var -> cmds_use_var var p1 || cmds_use_var var p2))
     <+>
-    (match p1 with [] -> Iter.empty | c1::c1s -> Iter.return (seq@[c1],c1s,p2))
+    map (fun p1 -> (seq,p1,p2)) (shrink_cmd_list_spine p1 ctx_doesnt_use_var)
     <+>
-    (match p2 with [] -> Iter.empty | c2::c2s -> Iter.return (seq@[c2],p1,c2s))
+    map (fun p2 -> (seq,p1,p2)) (shrink_cmd_list_spine p2 ctx_doesnt_use_var)
     <+>
-    concat_map (fun p1 -> Iter.map (fun p1 -> (seq,p1,p2)) (fix_cmds seq_env p1)) (Shrink.list_spine p1)
+    (fun yield -> (match p1 with [] -> Iter.empty | c1::c1s -> Iter.return (seq@[c1],c1s,p2)) yield)
```

Comment:
On that line, I’m not sure I see the point of η-expansion: it’s just delaying the `match` and `@` which will happen anyway, isn’t it? I would just drop it in the name of simpler code.

Reply:
It is indeed delaying the `match` and `@` computation :+1:
OCaml evaluates its arguments eagerly - also for the infix `<+>` parameters.
Omitting the eta-expansion adds needless work to the shrinker runs that successfully reduce before getting to these lines:
```ocaml
# Iter.find (fun _ -> true) Iter.(of_list [0;1] <+> (Printf.printf "boom!\n%!"; of_list [2;3]));;
boom!
- : int option = Some 0
# Iter.find (fun _ -> true) Iter.(of_list [0;1] <+> fun yield -> (Printf.printf "boom!\n%!"; of_list [2;3] yield));;
- : int option = Some 0
```
I agree that the code would read simpler without it though. I suggest we add a comment to explain why it is there.

Reply:
I've added a comment in 94649b9

Reply:
Ah! Indeed, shrinking will drop the tail as soon as one element does succeed. On first reading, I assumed we’ll eventually get to that item anyway. Thanks for the clarification.